include/sunmatrix/sunmatrix_ginkgoblock.hpp

```diff
@@ -0,0 +1,399 @@
+  BlockMatrix(gko::size_type num_blocks, sunindextype M, sunindextype N,
+              sunindextype num_nonzeros,
+              std::shared_ptr<const gko::Executor> gko_exec, SUNContext sunctx);
+
+  BlockMatrix(std::shared_ptr<GkoBatchMatType> gko_mat, SUNContext sunctx)
+    : sundials::impl::BaseMatrix(sunctx), gkomtx_(gko_mat)
+  {
+    initSUNMatrix();
+  }
+
+  // Move constructor
+  BlockMatrix(BlockMatrix&& that_matrix) noexcept
+    : sundials::impl::BaseMatrix(std::forward<BlockMatrix>(that_matrix)),
+      gkomtx_(std::move(that_matrix.gkomtx_))
+  {}
+
+  // Copy constructor clones the gko::matrix and SUNMatrix
+  BlockMatrix(const BlockMatrix& that_matrix)
+    : sundials::impl::BaseMatrix(that_matrix),
+      gkomtx_(gko::clone(that_matrix.gkomtx_))
+  {}
+
+  // Move assignment
+  BlockMatrix& operator=(BlockMatrix&& rhs) noexcept
+  {
+    gkomtx_ = std::move(rhs.gkomtx_);
+    sundials::impl::BaseMatrix::operator=(std::forward<BlockMatrix>(rhs));
+    return *this;
+  }
+
+  // Copy assignment clones the gko::matrix and SUNMatrix
+  BlockMatrix& operator=(const BlockMatrix& rhs)
+  {
+    gkomtx_ = gko::clone(rhs.gkomtx_);
+    sundials::impl::BaseMatrix::operator=(rhs);
+    return *this;
+  }
+
+  // Default destructor is fine since all members are RAII
+  virtual ~BlockMatrix() = default;
+
+  // Getters
+  std::shared_ptr<GkoBatchMatType> GkoMtx() const { return gkomtx_; }
+
+  std::shared_ptr<const gko::Executor> GkoExec() const
+  {
+    return GkoMtx()->get_executor();
+  }
+
+  const gko::batch_dim<2>& GkoSize() const { return GkoMtx()->get_size(); }
+
+  gko::size_type NumBlocks() const { return GkoSize().get_num_batch_items(); }
+
+  using sundials::impl::BaseMatrix::sunctx;
+
+  // Override the ConvertibleTo methods
+  operator SUNMatrix() override { return object_.get(); }
+
+  operator SUNMatrix() const override { return object_.get(); }
+
+  SUNMatrix Convert() override { return object_.get(); }
+
+  SUNMatrix Convert() const override { return object_.get(); }
+
+private:
+  std::shared_ptr<GkoBatchMatType> gkomtx_;
+
+  void initSUNMatrix()
+  {
+    this->object_->content = this;
+
+    this->object_->ops->getid = impl::SUNMatGetID_GinkgoBlock<GkoBatchMatType>;
+    this->object_->ops->clone = impl::SUNMatClone_GinkgoBlock<GkoBatchMatType>;
+    this->object_->ops->copy  = impl::SUNMatCopy_GinkgoBlock<GkoBatchMatType>;
+    // Ginkgo does not provide what we need for ScaleAdd except with BatchDense
+    this->object_->ops->scaleadd =
+      impl::SUNMatScaleAdd_GinkgoBlock<GkoBatchMatType>;
+    this->object_->ops->scaleaddi =
+      impl::SUNMatScaleAddI_GinkgoBlock<GkoBatchMatType>;
+    this->object_->ops->matvec = impl::SUNMatMatvec_GinkgoBlock<GkoBatchMatType>;
+    this->object_->ops->destroy = impl::SUNMatDestroy_GinkgoBlock<GkoBatchMatType>;
+  }
+};
+
+//
+// Class method specializations.
+//
+
+template<>
+inline BlockMatrix<GkoBatchDenseMat>::BlockMatrix(
```

Comment:
I don't understand what's going on here even with the comment above about not supporting general `GkoBatchMatType`. A bit more of an explanation in the comments would be helpful.